New style GEDelectron and GEDphoton regressions to EGM reco sequence in 90X

rafaellopesdesa opened the pull request:
PR for new style of EGM energy correction and per-particle resolution regression. It is similar to PR #16968 which was intended for standalone application by users (and that is still under review). This one modifies the EGM reco sequence, so changes are expected in the validation plots, specially in plots related to electron and photon energy. This PR depends on the PR #17048, which I believe has already been merged, with the new name convention for the regressions in the GlobalTag. I've tested locally in details with workflow 250200.0 and the results were good.

Att: @fcouderc

rafaellopesdesa committed:
add new regression to EGM reco sequence in 90X, requires PR 17048

cmsbuild commented:
A new Pull Request was created by  @rafaellopesdesa (Rafael Lopes de Sa) for CMSSW_9_0_X.

It involves the following packages:

RecoEgamma/EgammaTools


@cmsbuild, @cvuosalo, @slava77, @davidlange6 can you please review it and eventually sign? Thanks.
@Sam-Harper, @lgray this is something you requested to watch as well.
@davidlange6, @smuzaffar you are the release manager for this.

cms-bot commands are listed here cms-sw/cmssw#13028

slava77 commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

slava77 commented on RecoEgamma/EgammaTools/plugins/EGExtraInfoModifierFromDB.cc:
```diff
@@ -82,26 +85,31 @@ class EGExtraInfoModifierFromDB : public ModifyObjectValueBase {
+
+  bool useLocalFile_;
+  std::string addressLocalFile_;  
+  TFile* pointerLocalFile_;
```
additional open files during execution should be avoided as much as possible.
Open the file, read what's needed from it into memory and then close the file.
Do this if memory profile allows.

slava77 commented on RecoEgamma/EgammaTools/plugins/EGExtraInfoModifierFromDB.cc:
```diff
@@ -111,22 +119,20 @@ DEFINE_EDM_PLUGIN(ModifyObjectValueFactory,
+  useLocalFile_     = conf.getParameter<bool>("useLocalFile");
+  if (useLocalFile_) {
+    addressLocalFile_ = conf.getParameter<std::string>("addressLocalFile");
+    pointerLocalFile_ = TFile::Open(addressLocalFile_.c_str());
```
use FileInPath

slava77 commented on RecoEgamma/EgammaTools/plugins/EGExtraInfoModifierFromDB.cc:
```diff
@@ -364,118 +399,126 @@ namespace {
+  if ( !ecalRecHits.isValid() ) return;
   
-  std::array<float, 33> eval;  
+  Int_t N_SATURATEDXTALS  = 0;
```
please use base C++ types.
Also, all caps for a variable is a bad style, it's usually used for compiler macros and some constants.

slava77 commented on RecoEgamma/EgammaTools/plugins/EGExtraInfoModifierFromDB.cc:
```diff
@@ -364,118 +399,126 @@ namespace {
+  eval[0]  = raw_energy;
+  eval[1]  = the_sc->etaWidth();
+  eval[2]  = the_sc->phiWidth(); 
+  eval[3]  = full5x5_ess.e5x5/raw_energy;
```
check for full5x5_ess.e5x5 to be non-zero and set all values with potential divisions by zero to something sensible below. (maybe define ``` e5x5Inverse = full5x5_ess.e5x5 != 0 ? 1.f/full5x5_ess.e5x5 : 0 ``` )

slava77 replied:
Please define ``` rawInverse ``` in  a similar way. It's used multiple times here (in [3] and then in [25-27])

slava77 commented on RecoEgamma/EgammaTools/plugins/EGExtraInfoModifierFromDB.cc:
```diff
@@ -484,189 +527,236 @@ void EGExtraInfoModifierFromDB::modifyObject(reco::GsfElectron& ele) const {
+  float ptMode       = el_track->ptMode();
+  float ptModeErrror = el_track->ptModeError();
+  float etaModeError = el_track->etaModeError();
+  float pModeError   = sqrt(ptModeErrror*ptModeErrror*cosh(trkEta)*cosh(trkEta) + ptMode*ptMode*sinh(trkEta)*sinh(trkEta)*etaModeError*etaModeError);
```
Can you use qoverpModeError instead of this rather unnatural value?

slava77 commented on RecoEgamma/EgammaTools/python/regressionWeights_cfi.py:
```diff
@@ -1,29 +1,82 @@
-from CondCore.DBCommon.CondDBSetup_cfi import *
+import FWCore.ParameterSet.Config as cms
+
+def regressionWeights(process):
```
why is this file needed if you already have the GT updated in the release?

cmsbuild commented:
-1

Tested at: ea7d9b8fa00dcd3190cca74586b878a041d28892

You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-17101/17165/summary.html

I found follow errors while testing this PR

**Failed tests:  RelVals  AddOn**

* **RelVals**:

When I ran the RelVals I found an error in the following worklfows:
5.1 step1
<pre>runTheMatrix-results/5.1_TTbar+TTbarFS+HARVESTFS/step1_TTbar+TTbarFS+HARVESTFS.log

</pre>9.0 step3
<pre>runTheMatrix-results/9.0_Higgs200ChargedTaus+Higgs200ChargedTaus+DIGI+RECO+HARVEST/step3_Higgs200ChargedTaus+Higgs200ChargedTaus+DIGI+RECO+HARVEST.log

</pre>25.0 step3
<pre>runTheMatrix-results/25.0_TTbar+TTbar+DIGI+RECOAlCaCalo+HARVEST+ALCATT/step3_TTbar+TTbar+DIGI+RECOAlCaCalo+HARVEST+ALCATT.log

</pre>

* **AddOn**:

I found errors in the following addon tests:

cmsDriver.py TTbar_8TeV_TuneCUETP8M1_cfi  --conditions auto:run1_mc --fast  -n 100 --eventcontent AODSIM,DQM --relval 100000,1000 -s GEN,SIM,RECOBEFMIX,DIGI:pdigi_valid,L1,DIGI2RAW,L1Reco,RECO,EI,HLT:@fake,VALIDATION                 --customise=HLTrigger/Configuration/CustomConfigs.L1THLT --datatier GEN-SIM-DIGI-RECO,DQMIO --beamspot Realistic8TeVCollision : FAILED - time: date Wed Dec 28 20:36:01 2016-date Wed Dec 28 20:33:45 2016 s - exit: 16640
cmsDriver.py RelVal -s HLT:Fake,RAW2DIGI,L1Reco,RECO --mc --scenario=pp -n 10 --conditions auto:run1_mc_Fake --relval 9000,50 --datatier "RAW-HLT-RECO" --eventcontent FEVTDEBUGHLT --customise=HLTrigger/Configuration/CustomConfigs.L1THLT --processName=HLTRECO --filein file:RelVal_Raw_Fake_MC.root --fileout file:RelVal_Raw_Fake_MC_HLT_RECO.root : FAILED - time: date Wed Dec 28 20:40:28 2016-date Wed Dec 28 20:34:11 2016 s - exit: 16640

cmsbuild commented:
Comparison not run due to runTheMatrix errors (RelVals and Igprof tests were also skipped)

rafaellopesdesa commented:
@slava77 I will apply the corrections in the code review. I have some comments about the other questions here:

   * I tried to run the workflows that failed locally. They failed because the regressions cannot be found in the mcRun1 GlobalTags (I ran adding the regressions by hand). I have contacted the AlCa conveners to check why the mcRun2 queues were updated but not the mcRun1 ones.
   * Yes, we could have used a more natural definition of the uncertainty in the track momentum. In particular error on qOverp is an excellent suggestion. However, the training has already be done with this (rather convoluted, I agree) definition. We will keep your suggestion in mind for the next round.
   * regressionWeights_cfi.py is not necessary. I just updated it for "historical reasons". It was there before and I updated it pointing to the new names (which supposedly are in the GT). If preferred, I can remove this configuration from RecoEgamma.

Should I go updating the code with your review or should I wait the confirmation from AlCa that the mcRun1 queues have been committed as well before doing that?

slava77 commented:
On 12/28/16 12:43 PM, Rafael Lopes de Sa wrote:
> @slava77 <https://github.com/slava77> I will apply the corrections in
> the code review. I have some comments about the other questions here:
> 
>   * I tried to run the workflows that failed locally. They failed
>     because the regressions cannot be found in the mcRun1 GlobalTags (I
>     ran adding the regressions by hand). I have contacted the AlCa
>     conveners to check why the mcRun2 queues were updated but not the
>     mcRun1 ones.

OK.
Assuming the payloads are available, it should be straightforward to get
the updates in.

>   * Yes, we could have used a more natural definition of the uncertainty
>     in the track momentum. In particular error on qOverp is an excellent
>     suggestion. However, the training has already be done with this
>     (rather convoluted, I agree) definition. We will keep your
>     suggestion in mind for the next round.

Can you check the math with a few printouts for what goes into the
training inputs
using qOverP and this pError.
It could be that the final value that's computed will be the same, to
within numerical precision.
The problem is that you are making a numerical roundtrip from qOverP to
pt and eta and then back to p, which is slow and can be a source of bad
values for no good reason.


>   * regressionWeights_cfi.py is not necessary. I just updated it for
>     "historical reasons". It was there before and I updated it pointing
>     to the new names (which supposedly are in the GT). If preferred, I
>     can remove this configuration from RecoEgamma.

Please remove it.

> 
> Should I go updating the code with your review or should I wait the
> confirmation from AlCa that the mcRun1 queues have been committed as
> well before doing that?

Both will be necessary to proceed with this PR.
It's up to you to wait or update the code now.


Is it correct to assume that you will not need to support the old
training with the old variables in 90X?

Do we now have a copy of the new training processing code here and in
EgammaAnalysis?
If so, the copy in the analysis better be removed to reduce possible
confusion when someone just keeps using 80X recipes on 91X.

> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cms-sw/cmssw/pull/17101#issuecomment-269538062>, or
> mute the thread
> <https://github.com/notifications/unsubscribe-auth/AEdcbser0grC3KOXRQVcakW-MHh6UhF7ks5rMsoOgaJpZM4LXHH4>.
>

rafaellopesdesa committed:
code review modifications

cmsbuild commented:
Pull request #17101 was updated. @cmsbuild, @cvuosalo, @slava77, @davidlange6 can you please check and sign again.

rafaellopesdesa commented:
Hi @slava77,

I've updated the codes with your recommendations. Some further comments below:

   * Yes, using the qOverp error returned the same numerical value (to numerical precision). I changed as you recommended.
   * You are correct that the old regression style will *not* be supported in 90X.
   * No, we don't want the EgammaAnalysis code in 90X. I agree with you that this would be useless code repetition. I prefer to have the analysis-level application code just in 80X. If you want, I can port the code recommendation made in this PR to #16968
   * I tested other run1 workflows, and all seem to fail because of the lack of the payloads in the GT... as I mentioned above, I wrote to @mmusich about this issue. When the GTs are update we proceed with the tests here. :)

Thanks.

rafaellopesdesa committed:
code revision... again

cmsbuild commented:
Pull request #17101 was updated. @cmsbuild, @cvuosalo, @slava77, @davidlange6 can you please check and sign again.

slava77 commented:
Hi Rafael

On 12/28/16 2:07 PM, Rafael Lopes de Sa wrote:
> Hi @slava77 <https://github.com/slava77>,
> 
> I've updated the codes with your recommendations. Some further comments
> below:
> 
>   * Yes, using the qOverp error returned the same numerical value (to
>     numerical precision). I changed as you recommended.

Nice.

>   * You are correct that the old regression style will /not/ be
>     supported in 90X.
>   * No, we don't want the EgammaAnalysis code in 90X. I agree with you
>     that this would be useless code repetition. I prefer to have the
>     analysis-level application code just in 80X. If you want, I can port
>     the code recommendation made in this PR to #16968
>     <https://github.com/cms-sw/cmssw/pull/16968>

I thought that EGExtraInfoModifierFromDBUser.cc was proposed already for 90X
when I wrote the last message.
Good that it wasn't.

Which part of the recipe did you have in mind to port from (the yet
unmerged) 80X PR?
Is it going to be the code to possibly read (mini)AOD made with the old
regressions in 80X
to modify/update using the new one and use them in 90X?
This doesn't seem like a necessary use case.


>   * I tested other run1 workflows, and all seem to fail because of the
>     lack of the payloads in the GT... as I mentioned above, I wrote to
>     @mmusich <https://github.com/mmusich> about this issue. When the GTs
>     are update we proceed with the tests here. :)

OK


> 
> Thanks.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cms-sw/cmssw/pull/17101#issuecomment-269548805>, or
> mute the thread
> <https://github.com/notifications/unsubscribe-auth/AEdcbk2VWgDGn-L6FpPXmyOBvhHlQR2Wks5rMt2lgaJpZM4LXHH4>.
>

rafaellopesdesa commented:
Hi Slava,

Answering specifically the new questions:

   * Which part of the recipe did you have in mind to port from (the yet
unmerged) 80X PR?

Not from, but to. I meant recommendations concerning variable names and types, recipe to calculate pModeError, and so on... I am really interested in having the 80X PR merged. Even though it is an analysis-level PR, a lot of people are already applying the energy corrections in their analysis and I think it's much cleaner to have it in the central repository than in my personal area (that it naturally getting more and more outdated with every merge in 80X).

   * Is it going to be the code to possibly read (mini)AOD made with the old
regressions in 80X to modify/update using the new one and use them in 90X?

First, let me second your opinion that I don't think this will be necessary, but...

with this PR, what you said can be done at configuration level in RecoEgamma (I couldn't do this in 80X because the release was already frozen to anything changing RECO). I could provide a configuration suited to re-run (mini)AOD and add the new regression. Even better, I would do that as an option in regressionModifier_cfi.py. 

But, for the meantime, I won't add this option because I honestly think no one will need it at this point.

Thanks.

slava77 commented on RecoEgamma/EgammaTools/plugins/EGExtraInfoModifierFromDB.cc:
```diff
@@ -253,52 +259,83 @@ void EGExtraInfoModifierFromDB::setEvent(const edm::Event& evt) {
-  const std::vector<std::string> ph_condnames_mean  = (bunchspacing_ == 25) ? ph_conf.condnames_mean_25ns  : ph_conf.condnames_mean_50ns;
-  const std::vector<std::string> ph_condnames_sigma = (bunchspacing_ == 25) ? ph_conf.condnames_sigma_25ns : ph_conf.condnames_sigma_50ns;
+  if (useLocalFile_)
+    pointerLocalFile_ = TFile::Open(addressLocalFile_.fullPath().c_str(), "READ");
```
file open and close every event is another no-go.
Please move to the constructor or something that's not executed every event.

slava77 commented:
On 12/28/16 3:33 PM, Rafael Lopes de Sa wrote:
> Hi Slava,
> 
> Answering specifically the new questions:
> 
>   * Which part of the recipe did you have in mind to port from (the yet
>     unmerged) 80X PR?
> 
> Not from, but to.
> 

Ah, sorry.


The 80X PR is a sideways "backport" with  different reviewers assigned.
Not particularly close to ideal situation.

It may be more practical that the new
EGExtraInfoModifierFromDB is renamed to EGExtraInfoModifierFromDB2016
(or a similar name).
It will then be kept in RecoEgamma/EgammaTools in both 80X and 90X.
(the old EGExtraInfoModifierFromDB can be removed already in this PR).
The backport to 80X then becomes more straightforward.

rafaellopesdesa committed:
only open file once in debug mode and do not fill vector of GBRForests repeatedly in normal operations

rafaellopesdesa committed:
only open file once in debug mode and do not fill vector of GBRForests repeatedly in normal operations

cmsbuild commented:
Pull request #17101 was updated. @cmsbuild, @cvuosalo, @slava77, @davidlange6 can you please check and sign again.

rafaellopesdesa commented:
Hi Slava,

Ok, I updated the code to just open the file once. I also notice that the original code was growing the vector of GBRForestD pointers in every event. Now I just reserve the size I need and update in every event.

Rafael

rafaellopesdesa commented:
About the suggestion of how to deal with the 80X PR, I think I prefer the file structure as it is right now. I understand that is less ideal for code review, but I think it is easier for us in EGM to find things in the future. As I've explained in the 80X PR, EgammaAnalysis has been traditionally used for user-level applications, and I was hoping to be faithful to this idea.

Let's see how these PR go. Tomorrow I will update the 80X PR with the improvements discussed today in this thread.

mmusich commented:
@rafaellopesdesa @slava77
about:

>  I tested other run1 workflows, and all seem to fail because of the lack of the payloads in the GT... as I mentioned above, I wrote to @mmusich about this issue. When the GTs are update we proceed with the tests here. :)

the updates are being tested now in #17104. I suggest that commit https://github.com/cms-AlCaDB/cmssw/commit/aee8c69da128c444aca18ce03ee1d8297c94baaf is pulled in together with the rest of the code changes requested during the review to be able to execute tests. 
Cheers,

M.

slava77 commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 4b9089e8c0379df2cf4b7b47eddc0a8d78281569
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-17101/17219/summary.html

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-17101/17219/summary.html

slava77 commented:
@rafaellopesdesa 
I was looking at slides in the EGM/ECAL meeting today.
Should I interpret that you will soon update the payloads or/and the variables used in training?
Please clarify.
Thank you.

slava77 commented on RecoEgamma/EgammaTools/plugins/EGExtraInfoModifierFromDB.cc:
```diff
@@ -253,52 +306,62 @@ void EGExtraInfoModifierFromDB::setEvent(const edm::Event& evt) {
 
-  const std::vector<std::string> ph_condnames_mean  = (bunchspacing_ == 25) ? ph_conf.condnames_mean_25ns  : ph_conf.condnames_mean_50ns;
-  const std::vector<std::string> ph_condnames_sigma = (bunchspacing_ == 25) ? ph_conf.condnames_sigma_25ns : ph_conf.condnames_sigma_50ns;
+  const std::vector<std::string> ph_condnames_ecalonly_mean  = ph_conf.condnames_ecalonly_mean;
```
use ```const std::vector<std::string>& ph_condnames_ecalonly_mean  = ph_conf.condnames_ecalonly_mean;``` to avoid unnecessary copies

slava77 replied:
same for all other instances in this file

slava77 commented on RecoEgamma/EgammaTools/plugins/EGExtraInfoModifierFromDB.cc:
```diff
@@ -364,118 +426,128 @@ namespace {
-  std::array<float, 33> eval;  
+  int nSaturatedXtals  = 0;
+  std::vector< std::pair<DetId, float> > hitsAndFractions = theseed->hitsAndFractions();
+  for (auto hitFractionPair : hitsAndFractions) {    
```
```const std::vector< std::pair<DetId, float> >& hitsAndFractions ```
``` for (auto const& hitFractionPair : hitsAndFractions ```

slava77 commented on RecoEgamma/EgammaTools/plugins/EGExtraInfoModifierFromDB.cc:
```diff
@@ -364,118 +426,128 @@ namespace {
-    
-    if(theseed == pclus ) 
+  size_t i_cluster = 0;
+  for( auto clus = the_sc->clustersBegin(); clus != the_sc->clustersEnd(); ++clus ) {
```
why not just ``` for (auto const& pclus: the_sc->clusters() ) ```  ?

slava77 commented on RecoEgamma/EgammaTools/plugins/EGExtraInfoModifierFromDB.cc:
```diff
@@ -364,118 +426,128 @@ namespace {
   //(These should be stored inside the conditions object in the future as well)
-  constexpr double meanlimlow  = 0.2;
-  constexpr double meanlimhigh = 2.0;
+  constexpr double meanlimlow  = -1.0;
```
Looks like the comment about saving these magic numbers in the conditions did not serve the purpose.
Better happen next time.

slava77 commented on RecoEgamma/EgammaTools/plugins/EGExtraInfoModifierFromDB.cc:
```diff
@@ -484,189 +556,234 @@ void EGExtraInfoModifierFromDB::modifyObject(reco::GsfElectron& ele) const {
+
+  int nSaturatedXtals  = 0;
+  std::vector< std::pair<DetId, float> > hitsAndFractions = theseed->hitsAndFractions();
+  for (auto hitFractionPair : hitsAndFractions) {
```
as above, add ```& ``` to avoid copies

slava77 commented on RecoEgamma/EgammaTools/plugins/EGExtraInfoModifierFromDB.cc:
```diff
@@ -484,189 +556,234 @@ void EGExtraInfoModifierFromDB::modifyObject(reco::GsfElectron& ele) const {
+  edm::Ptr<reco::CaloCluster> pclus;
+  // loop over all clusters that aren't the seed  
+  size_t i_cluster = 0;
+  for( auto clus = the_sc->clustersBegin(); clus != the_sc->clustersEnd(); ++clus ) {
```
as above, range loop can be simpler

rafaellopesdesa commented:
Hi Slava,

Yes, Fabrice pointed to me an inconsistency between the training code and
application code, so a small update will be introduced in the coming days.
I will also reply to your more recent code review concomitantly.

Thanks again,
-- Rafael.

On Thu, Jan 12, 2017 at 5:12 AM, Slava Krutelyov <notifications@github.com>
wrote:

> *@slava77* commented on this pull request.
> ------------------------------
>
> In RecoEgamma/EgammaTools/plugins/EGExtraInfoModifierFromDB.cc
> <https://github.com/cms-sw/cmssw/pull/17101#pullrequestreview-16225105>:
>
> > +  eval[22]  = full5x5_pss.e2x5Bottom*e5x5Inverse;
> +  eval[23]  = nSaturatedXtals;
> +  eval[24]  = std::max(0,numberOfClusters);
> +
> +  // calculate sub-cluster variables
> +  std::vector<float> clusterRawEnergy;
> +  clusterRawEnergy.resize(std::max(3, numberOfClusters), 0);
> +  std::vector<float> clusterDEtaToSeed;
> +  clusterDEtaToSeed.resize(std::max(3, numberOfClusters), 0);
> +  std::vector<float> clusterDPhiToSeed;
> +  clusterDPhiToSeed.resize(std::max(3, numberOfClusters), 0);
> +
> +  edm::Ptr<reco::CaloCluster> pclus;
> +  // loop over all clusters that aren't the seed
> +  size_t i_cluster = 0;
> +  for( auto clus = the_sc->clustersBegin(); clus != the_sc->clustersEnd(); ++clus ) {
>
> as above, range loop can be simpler
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cms-sw/cmssw/pull/17101#pullrequestreview-16225105>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFtCpHcPEK4re83Ripiq29rG-sVZSM4oks5rRTfKgaJpZM4LXHH4>
> .
>

slava77 commented:
@rafaellopesdesa 
I just wanted to check the progress here.

slava77 commented:
(almost) weekly ping.
@rafaellopesdesa 
Please update on the status of this PR

Thank you.

rafaellopesdesa commented:
@slava77 Iterating with AlCA to add the payloads and just made an associated PR so that I don't have to re-calculate values when applying the regression.

rafaellopesdesa commented:
Closing and merging with #17370 with the corrections required in both reviews.
